[wasm][xharness] install development SSL certificate on Helix agent before xharness run via SDK or powershell

pavelsavara opened the pull request:
Caused by https://github.com/dotnet/runtime/pull/53180, https://github.com/dotnet/runtime/pull/53225
- The original approach to install certificates didn't work because we don't have dotnet SDK, just runtime on Helix agents.
- `dotnet dev-certs https` needs user interaction with a dialog. Instead, using powershelgl to install certs
- `Wasm.Build.Tests` are being disabled here because of https://github.com/dotnet/runtime/issues/53405

Fixes https://github.com/dotnet/runtime/issues/53207

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

premun commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -98,21 +98,24 @@
-      Below is alternative to `dotnet dev-certs https` but we don't have full SDK installed on helix, so the tool is not available.
     -->
-    <HelixPreCommand Include="powershell -command &quot;New-SelfSignedCertificate -FriendlyName &#39;ASP.NET Core HTTPS development certificate&#39; -DnsName @(&#39;localhost&#39;) -Subject &#39;CN = localhost&#39; -KeyAlgorithm RSA -KeyLength 2048 -HashAlgorithm sha256 -CertStoreLocation &#39;Cert:\CurrentUser\My&#39; -TextExtension @(&#39;2.5.29.37={text}1.3.6.1.5.5.7.3.1&#39;,&#39;1.3.6.1.4.1. ...
+    <IncludeDotNetCli>true</IncludeDotNetCli>
```
This one and `DotNetCliPackageType` should go to a `<PropertyGroup>`

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

radical commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -163,14 +161,19 @@
       so if we don't use "call", then we cause the parent script to exit, and anything after will not be executed.
     -->
-    <HelixCommand Condition="'$(TargetsWindows)' == 'true' or '$(BrowserHost)' == 'windows'">call RunTests.cmd</HelixCommand>
+    <HelixCommand Condition="'$(TargetsWindows)' == 'true' or '$(BrowserHost)' == 'windows'">dotnet dev-certs https &amp;&amp; call RunTests.cmd</HelixCommand>
```
Any reason for not having `dotnet dev-certs https` as a `HelixPreCommand`?

pavelsavara replied:
See here https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-53277-merge-5628afbfb7314eaeae/System.Net.Http.Json.Functional.Tests/console.c58f5af9.log?sv=2019-07-07&se=2021-06-15T11%3A59%3A29Z&sr=c&sp=rl&sig=KW9SsPKDADoV8OokbFyjERkqajhQNEhFPrBxLk2gfw4%3D

It is inserted before PATH is set to dotnet SDK

radical replied:
hm then I would set `<HelixCommand>dotnet dev-certs https &amp;&amp;</HelixCommand`>, and the rest can append to it.

radical commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -98,22 +98,20 @@
     <HelixPreCommand Condition="'$(Scenario)' == 'WasmTestOnBrowser'" Include="export XHARNESS_COMMAND=test-browser" />
     <HelixPreCommand Include="export XHARNESS_DISABLE_COLORED_OUTPUT=true" />
     <HelixPreCommand Include="export XHARNESS_LOG_WITH_TIMESTAMPS=true" />
+    <HelixPreCommand Include="export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1" />
```
Just curious, why is this needed here?

pavelsavara replied:
First time experience does the cert installation too, plus some text dump. So it's not necessary.

premun replied:
It might be better to add it here: https://github.com/dotnet/arcade/blob/c336c9622eafce2b9bf3c4c164a4c0b57703aa5e/src/Microsoft.DotNet.Helix/Sdk/tools/dotnet-cli/DotNetCli.targets#L18

pavelsavara replied:
but arcade is not sure we have SDK, so it would have to be conditional

pavelsavara replied:
runtime doesn't have the cert tool only SDK does. you can't `dotnet dev-certs https` on runtime without failure.
we could remove both DOTNET_SKIP_FIRST_TIME_EXPERIENCE and `dotnet dev-certs https` from here and the cert would be installed on SDK first run. But it would be implicit side effect, rather than explicit action. Is that OK ?

pavelsavara replied:
$(TestPackages) is also setting DOTNET_SKIP_FIRST_TIME_EXPERIENCE, so we would have to avoid clash.

premun replied:
I meant, add the cert stuff in runtime as you plan to.

I was only talking about `DOTNET_SKIP_FIRST_TIME_EXPERIENCE` that it could go into Arcade too

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

radical commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -98,22 +98,26 @@
     -->
-    <HelixPreCommand Include="powershell -command &quot;New-SelfSignedCertificate -FriendlyName &#39;ASP.NET Core HTTPS development certificate&#39; -DnsName @(&#39;localhost&#39;) -Subject &#39;CN = localhost&#39; -KeyAlgorithm RSA -KeyLength 2048 -HashAlgorithm sha256 -CertStoreLocation &#39;Cert:\CurrentUser\My&#39; -TextExtension @(&#39;2.5.29.37={text}1.3.6.1.5.5.7.3.1&#39;,&#39;1.3.6.1.4.1. ...
-  </ItemGroup>
+    <_PreHelixCommand>dotnet dev-certs https &amp;&amp; <_PreHelixCommand/>
```
This is being used in only one specific case, and should be moved next to it.
```xml
<HelixCommand Condition="'$(TargetOS)' == 'Browser'">dotnet dev-certs https &amp;&amp; </HelixCommand>

<HelixCommand Condition="'$(TargetsWindows)' == 'true' or '$(BrowserHost)' == 'windows'">call RunTests.cmd</HelixCommand>
...
```

pavelsavara replied:
done

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

radical commented:
Browser windows:
```
C:\helix\work\workitem>dotnet dev-certs https   && call RunTests.cmd 

...
[EXECUTION TIMED OUT]
Exit Code:-3Executor timed out after 1800 seconds and was killed

['Microsoft.Extensions.Configuration.EnvironmentVariables.Tests' WORK ITEM LOG TRUNCATED: Command timed out, and was killed]
```

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

premun commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -163,14 +157,23 @@
+      For the HTTPS endpoint we need development SSL certificate. For that we need SDK.
+      After HelixPreCommand in arcade set the PATH to dotnet, we could use `dotnet dev-certs https`.
+    -->
+    <IncludeDotNetCli Condition="'$(TargetOS)' == 'Browser' and '$(BrowserHost)' != 'windows'">true</IncludeDotNetCli>
```
You can apply condition to a whole `PropertyGroup` to reduce code duplication

ericstj commented:
Not sure if this PR will fix it, but noticing lots of failures of the form:
```
Connection refused [::ffff:127.0.0.1]:43699 (localhost:43699)
...
Starting ChromeDriver 84.0.4147.0 (ff714bcb10c3356b1b1b2819148507a34fa4b1f8-refs/heads/master@{#768968}) on port 43699
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
Connection refused [::ffff:127.0.0.1]:43699 (localhost:43699)
ChromeDriver was started successfully.
Connection refused [::ffff:127.0.0.1]:43699 (localhost:43699)
[04:59:16] crit: System.InvalidOperationException: Unable to configure HTTPS endpoint. No server certificate was specified, and the default developer certificate could not be found or is out of date.
                 To generate a developer certificate run 'dotnet dev-certs https'. To trust the certificate (Windows and macOS only) run 'dotnet dev-certs https --trust'.
                 For more information on configuring HTTPS see https://go.microsoft.com/fwlink/?linkid=848054.
                    at Microsoft.AspNetCore.Hosting.ListenOptionsHttpsExtensions.UseHttps(ListenOptions listenOptions, Action`1 configureOptions)
                    at Microsoft.AspNetCore.Hosting.ListenOptionsHttpsExtensions.UseHttps(ListenOptions listenOptions)
                    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.AddressesStrategy.BindAsync(AddressBindContext context)
                    at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindAsync(IEnumerable`1 listenOptions, AddressBindContext context)
                    at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.BindAsync(CancellationToken cancellationToken)
                    at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.StartAsync[TContext](IHttpApplication`1 application, CancellationToken cancellationToken)
                    at Microsoft.AspNetCore.Hosting.WebHost.StartAsync(CancellationToken cancellationToken)
                    at Microsoft.DotNet.XHarness.CLI.Commands.WebServer.Start(TestCommandArguments arguments, ILogger logger, Func`2 onConsoleConnected, CancellationToken token) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/WebServer.cs:line 71
                    at Microsoft.DotNet.XHarness.CLI.Commands.Wasm.WasmBrowserTestRunner.RunTestsWithWebDriver(DriverService driverService, IWebDriver driver) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/Browser/WasmBrowserTestRunner.cs:line 60
                    at Microsoft.DotNet.XHarness.CLI.Commands.Wasm.WasmTestBrowserCommand.InvokeInternal(ILogger logger) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/Browser/WasmTestBrowserCommand.cs:line 73
                    at Microsoft.DotNet.XHarness.Common.CLI.Commands.XHarnessCommand.Invoke(IEnumerable`1 arguments) in /_/src/Microsoft.DotNet.XHarness.Common/CLI/Commands/XHarnessCommand.cs:line 120
XHarness exit code: 71 (GENERAL_FAILURE)
```

Here: https://dev.azure.com/dnceng/public/_build/results?buildId=1158915&view=ms.vss-test-web.build-test-results-tab&runId=34978388&paneView=debug

https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-53212-merge-53c30b46d5f14dc0b0/XsltCompiler.Tests/console.a4a5589c.log?sv=2019-07-07&se=2021-06-16T04%3A55%3A12Z&sr=c&sp=rl&sig=ZTsowdh4FokJOmTDEzlrz04dVXOJpWfNSu0XOkzmOmA%3D
https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-53212-merge-53c30b46d5f14dc0b0/System.IO.Tests/console.72e792b5.log?sv=2019-07-07&se=2021-06-16T04%3A55%3A12Z&sr=c&sp=rl&sig=ZTsowdh4FokJOmTDEzlrz04dVXOJpWfNSu0XOkzmOmA%3D

radical commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -98,21 +98,15 @@
-    -->
-    <HelixPreCommand Include="powershell -command &quot;New-SelfSignedCertificate -FriendlyName &#39;ASP.NET Core HTTPS development certificate&#39; -DnsName @(&#39;localhost&#39;) -Subject &#39;CN = localhost&#39; -KeyAlgorithm RSA -KeyLength 2048 -HashAlgorithm sha256 -CertStoreLocation &#39;Cert:\CurrentUser\My&#39; -TextExtension @(&#39;2.5.29.37={text}1.3.6.1.5.5.7.3.1&#39;,&#39;1.3.6.1.4.1. ...
+    <!-- SSL cert for kestrel in the xharness process on windows -->
+    <HelixPreCommand Include="powershell -command &quot;New-SelfSignedCertificate -FriendlyName &#39;ASP.NET Core HTTPS development certificate&#39; -DnsName @(&#39;localhost&#39;) -Subject &#39;CN = localhost&#39; -KeyAlgorithm RSA -KeyLength 2048 -HashAlgorithm sha256 -CertStoreLocation &#39;Cert:\CurrentUser\My&#39; -TextExtension @(&#39;2.5.29.37={text}1.3.6.1.5.5.7.3.1&#39;,&#39;1.3.6.1.4.1. ...
```
Remove trailing `</ItemGroup>`, fixes the build.
```suggestion
    <HelixPreCommand Include="powershell -command &quot;New-SelfSignedCertificate -FriendlyName &#39;ASP.NET Core HTTPS development certificate&#39; -DnsName @(&#39;localhost&#39;) -Subject &#39;CN = localhost&#39; -KeyAlgorithm RSA -KeyLength 2048 -HashAlgorithm sha256 -CertStoreLocation &#39;Cert:\CurrentUser\My&#39; -TextExtension @(&#39;2.5.29.37={text}1.3.6.1.5.5.7.3.1&#39;,&#39;1.3.6.1.4.1.311.84.1.1={hex}02&#39;,&#39;2.5.29.19={text}&#39;) -KeyUsage DigitalSignature,KeyEncipherment&quot;" />
```

premun reviewed (approved):
I hope this doesn't bite anyone because there will be a different SDK/runtime but I guess we only use it to start XHarness, right?

pavelsavara commented:
InvalidArgument FriendlyName , windows.81.amd64.open.rt
https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-53280-merge-63600af09d334c4aa4/Common.Tests/console.901aae2d.log?sv=2019-07-07&se=2021-06-16T12%3A09%3A57Z&sr=c&sp=rl&sig=YrMinuy%2FLcDqTi35w%2BSbCh2mJpItGcjoiFYHwm4STr4%3D

pavelsavara committed:
install SDK so that we could use dev-certs tool

pavelsavara commented:
Last failure was in `Wasm.Build.Tests`

```
  Starting:    Wasm.Build.Tests (parallel test collections = on, max threads = 2)
    Wasm.Build.Tests.InvariantGlobalizationTests.RelinkingWithoutAOT [STARTING]
    Wasm.Build.Tests.MainWithArgsTests.TopLevelWithArgs [STARTING]
============== DELETING THE BUILDS =============
============== DELETING THE BUILDS =============
    Wasm.Build.Tests.InvariantGlobalizationTests.RelinkingWithoutAOT [FAIL]
      System.TypeInitializationException : The type initializer for 'Wasm.Build.Tests.BuildTestBase' threw an exception.
      ---- System.TypeLoadException : Could not load type 'System.IO.DirectoryInfo' from assembly 'System.Runtime, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
      Stack Trace:
        /_/src/tests/BuildWasmApps/Wasm.Build.Tests/BuildTestBase.cs(123,0): at Wasm.Build.Tests.BuildTestBase.ConfigWithAOTData(Boolean aot)
        /_/src/tests/BuildWasmApps/Wasm.Build.Tests/MainWithArgsTests.cs(22,0): at Wasm.Build.Tests.MainWithArgsTests.MainWithArgsTestData(Boolean aot, RunHost host)
```
https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-53280-merge-607faae452b9457aa6/Wasm.Build.Tests/console.370f3b8e.log?sv=2019-07-07&se=2021-06-16T14%3A34%3A05Z&sr=c&sp=rl&sig=Xyb%2BqNpGVBevJ95zRErg9Y0xjZku759MvDVGK2lJTD4%3D

radical commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -157,6 +149,22 @@
+    We also run some network (HTTP and HTTPS) tests that connect to this server.
+    In order to set up an HTTPS endpoint we need to install a development SSL certificate.
+  -->
+  <PropertyGroup Condition="'$(TargetOS)' == 'Browser' and '$(HelixCommand)' == '' and '$(TargetsWindows)' != 'true' and '$(BrowserHost)' != 'windows'">
```
This consolidation breaks the groupings for the properties that we have for the various cases, and makes the condition complicated. I would suggest:

1. add this
```xml
  <PropertyGroup Condition="'$(TargetOS)' == 'Browser' and '$(NeedsDotNetSdkWithCli)' == 'true'">
    <IncludeDotNetCli>true</IncludeDotNetCli>
    <DotNetCliPackageType>sdk</DotNetCliPackageType>
  </PropertyGroup>
```

2. and then set `NeedsDotNetSdkWithCli` wherever we have the `IncludeDotNetCli=true`, and `DotNetCliPackageType=sdk`.

That should retain the local grouping, and have consolidation around those properties.

radical replied:
And maybe skipping the sdk for `Scenario=BuildWasmApps` will help with the failures in that.

radical commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -157,6 +149,22 @@
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(TargetOS)' == 'Browser' and '$(HelixCommand)' == '' and ('$(TargetsWindows)' == 'true' or '$(BrowserHost)' == 'windows')">
+    <!-- install SSL cert via powershell on windows -->
+    <InstallCert>powershell -command &quot;New-SelfSignedCertificate -FriendlyName &#39;ASP.NET Core HTTPS development certificate&#39; -DnsName @(&#39;localhost&#39;) -Subject &#39;CN = localhost&#39; -KeyAlgorithm RSA -KeyLength 2048 -HashAlgorithm sha256 -CertStoreLocation &#39;Cert:\CurrentUser\My&#39; -TextExtension @(&#39;2.5.29.37={text}1.3.6.1.5.5.7.3.1&#39;,&#39;1.3.6.1.4.1.311.84.1.1={h ...
```
Just rename this to `HelixCommandPrefix`, and use when setting `HelixCommand`, as you were doing earlier.

radical committed:
Refactor a bit

radical committed:
fix typo

radical committed:
dev-certs not needed for wasm build tests

radical committed:
Use powershell to install certs on windows

.. because `dotnet dev-certs https` shows a dialog on windows, which
would require user interaction.

radical committed:
Disable wasm.build.tests failing due to https://github.com/dotnet/runtime/issues/53405

radical commented:
Installer test failure: `Bundle_Is_Extracted`

```
System.ComponentModel.Win32Exception : Text file busy

Stack trace
   at System.Diagnostics.Process.ForkAndExecProcess(String filename, String[] argv, String[] envp, String cwd, Boolean redirectStdin, Boolean redirectStdout, Boolean redirectStderr, Boolean setCredentials, UInt32 userId, UInt32 groupId, UInt32[] groups, Int32& stdinFd, Int32& stdoutFd, Int32& stderrFd, Boolean usesTerminal, Boolean throwOnNoExec)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at Microsoft.DotNet.Cli.Build.Framework.Command.Start() in /_/src/installer/tests/TestUtils/Command.cs:line 199
   at Microsoft.DotNet.Cli.Build.Framework.Command.Execute(Boolean fExpectedToFail) in /_/src/installer/tests/TestUtils/Command.cs:line 239
   at Microsoft.DotNet.Cli.Build.Framework.Command.Execute() in /_/src/installer/tests/TestUtils/Command.cs:line 171
   at AppHost.Bundle.Tests.NetCoreApp3CompatModeTests.Bundle_Is_Extracted() in /_/src/installer/tests/Microsoft.NET.HostModel.Tests/AppHost.Bundle.Tests/NetCoreApp3CompatModeTests.cs:line 33
```

Tracked at https://github.com/dotnet/runtime/issues/43316 . This seems to be an intermittent error, and unrelated to this PR. Merging.
